### PR TITLE
Always close _all_ `ijson` coroutines, even if doing so raises Exceptions

### LINKED ISSUE #13293: Close `ijson` coroutines ourselves instead of letting the GC close them
https://github.com/matrix-org/synapse/blob/d9f092285b28f0cdac0d985813a1cabd8ea990b6/synapse/federation/transport/client.py#L928-L930
https://github.com/matrix-org/synapse/blob/d9f092285b28f0cdac0d985813a1cabd8ea990b6/synapse/federation/transport/client.py#L972-L974

These `close()` calls were added in #12875, but I overlooked that we'd only close the first coroutine if it raised an `IncompleteJSONError`.
We need to close _all_ coroutines, even when any of them raise an error.

Until we fix this properly, we'll get unhelpful `Exception ignored in: <generator object utf8_encoder at 0x7f87c9b1ccf0>` errors in sentry without a stack trace or logging context.

### PR DESCRIPTION
**Pull request by DMRobertson**
Fixes #13293. Improves on #12875 and #9958.

This change means that we should raise one Sentry event for each such failure to `.finish()` a parser, instead of an event for each line printed to stderr.

I have only included the first traceback `exceptions[0]`: I'm assuming that's enough to diagnose what happened if we see this error. I do make sure to include the `str()` of each individual exception though.

I considered using https://pypi.org/project/exceptiongroup/ but this was a little scary: it mentions patching TracebackException and adding exception hooks; I feared that this wouldn't play nicely with twisted + logcontexts + other arcanery in Synapse I don't grok.

### CODE REVIEW

**Comment by squahtx** [comment]
Thanks for picking this up!

File: synapse/federation/transport/client.py [hunk 1]
```diff
@@ -970,6 +970,27 @@ def write(self, data: bytes) -> int:
...
+        try:
+            c.close()
+        except Exception as e:
+            exceptions.append(e)
+
+    if exceptions:
+        # raise from the first exception so that the traceback has slightly more context
+        raise ExceptionBundle(
+            f"There were {len(exceptions)} errors closing coroutines", exceptions
+        ) from exceptions[0]
```
Comment (DMRobertson) [inline synapse/federation/transport/client.py:996]: Not promising this is the cleanest way to do it---there are no tracebacks for `exceptions[1:]`---but this seemed Good Enough.

File: synapse/util/__init__.py [hunk 2]
```diff
@@ -193,3 +193,15 @@ def log_failure(
...
+
+class ExceptionBundle(Exception):
+    # A poor stand-in for something like Python 3.11's ExceptionGroup.
+    # (A backport called `exceptiongroup` exists but seems overkill: we just want a
+    # container type here.)
+    def __init__(self, message: str, exceptions: Sequence[Exception]):
+        parts = [message]
+        for e in exceptions:
+            parts.append(str(e))
+        super().__init__("\n  - ".join(parts))
```
Comment (DMRobertson) [inline synapse/util/__init__.py:206]: I'm specifically trying to ensure each exception's summary is included here, even if we don't have their tracebacks.

There are some comments in the description elaborating on why I didn't want to use the backport here,

File: synapse/federation/transport/client.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/federation/transport/client.py:996]: (I'm not sure what @richvdh meant by exception chanining in https://github.com/matrix-org/synapse/issues/13293#issuecomment-1187643041---something along these lines?)

File: tests/federation/transport/test_client.py [hunk 3]
```diff
@@ -94,3 +95,34 @@ def test_servers_in_room(self) -> None:
         # Retrieve and check the parsed SendJoinResponse
         parsed_response = parser.finish()
         self.assertEqual(parsed_response.servers_in_room, ["hs1", "hs2"])
+
+    def test_errors_closing_coroutines(self) -> None:
```
Comment (reivilibre) [inline tests/federation/transport/test_client.py:99]: docstring please, just so the context is kept around for years to come

Comment (DMRobertson) [reply tests/federation/transport/test_client.py:99]: Sure, this is 99a0ad8.

**Pull request merged**